### PR TITLE
avoid calls to make_current() and make_clear() by using asyncio.run in LoopRunner

### LINKED ISSUE #6784: make_current/clear_current are deprecated in Tornado
**What happened**:
`make_current` and `clear_current` appear to be deprecated in Tornado 6.2, which causes errors in tests, since warnings are treated as errors. Though it's my mistake that I left that on, it's probably something that should be fixed soonish.

```pytb
_______________________ test_loop_started_in_constructor _______________________

cleanup = None

    def test_loop_started_in_constructor(cleanup):
        # test that SpecCluster.__init__ starts a loop in another thread
>       cluster = SpecCluster(worker_spec, scheduler=scheduler, loop=None)

../../BUILDROOT/python-distributed-2022.7.1-1.fc37.x86_64/usr/lib/python3.11/site-packages/distributed/deploy/tests/test_spec_cluster.py:87: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
../../BUILDROOT/python-distributed-2022.7.1-1.fc37.x86_64/usr/lib/python3.11/site-packages/distributed/deploy/spec.py:255: in __init__
    super().__init__(
../../BUILDROOT/python-distributed-2022.7.1-1.fc37.x86_64/usr/lib/python3.11/site-packages/distributed/deploy/cluster.py:67: in __init__
    self._loop_runner = LoopRunner(loop=loop, asynchronous=asynchronous)
../../BUILDROOT/python-distributed-2022.7.1-1.fc37.x86_64/usr/lib/python3.11/site-packages/distributed/utils.py:440: in __init__
    self._loop = IOLoop()
/usr/lib64/python3.11/site-packages/tornado/util.py:276: in __new__
    instance.initialize(*args, **init_kwargs)
/usr/lib64/python3.11/site-packages/tornado/platform/asyncio.py:339: in initialize
    super().initialize(**kwargs)
/usr/lib64/python3.11/site-packages/tornado/platform/asyncio.py:140: in initialize
    super().initialize(**kwargs)
/usr/lib64/python3.11/site-packages/tornado/ioloop.py:350: in initialize
    self.make_current()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

self = <tornado.platform.asyncio.AsyncIOLoop object at 0x7fb8eff84690>

    def make_current(self) -> None:
>       warnings.warn(
            "make_current is deprecated; start the event loop first",
            DeprecationWarning,
            stacklevel=2,
        )
E       DeprecationWarning: make_current is deprecated; start the event loop first

/usr/lib64/python3.11/site-packages/tornado/platform/asyncio.py:353: DeprecationWarning
```

```pytb
_____________________ test_client_async_before_loop_starts _____________________

cleanup = None

    def test_client_async_before_loop_starts(cleanup):
        async def close():
            async with client:
                pass
    
>       with pristine_loop() as loop:

../../BUILDROOT/python-distributed-2022.7.1-1.fc37.x86_64/usr/lib/python3.11/site-packages/distributed/tests/test_client.py:5540: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
/usr/lib64/python3.11/contextlib.py:137: in __enter__
    return next(self.gen)
../../BUILDROOT/python-distributed-2022.7.1-1.fc37.x86_64/usr/lib/python3.11/site-packages/distributed/utils_test.py:222: in pristine_loop
    IOLoop.clear_instance()
/usr/lib64/python3.11/site-packages/tornado/ioloop.py:227: in clear_instance
    IOLoop.clear_current()
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 

    @staticmethod
    def clear_current() -> None:
        """Clears the `IOLoop` for the current thread.
    
        Intended primarily for use by test frameworks in between tests.
    
        .. versionchanged:: 5.0
           This method also clears the current `asyncio` event loop.
        .. deprecated:: 6.2
        """
>       warnings.warn(
            "clear_current is deprecated",
            DeprecationWarning,
            stacklevel=2,
        )
E       DeprecationWarning: clear_current is deprecated

/usr/lib64/python3.11/site-packages/tornado/ioloop.py:317: DeprecationWarning
```

**What you expected to happen**:
Tests pass.

**Environment**:

- Dask version: 2022.7.1
- Python version: 3.11.0.b4
- Operating System: Fedora Rawhide
- Install method (conda, pip, source): source

### PR DESCRIPTION
**Pull request by graingert**
Closes #6784

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

File: distributed/deploy/cluster.py [hunk 1]
```diff
@@ -521,7 +523,8 @@ def __enter__(self):
         return self.sync(self.__aenter__)
 
     def __exit__(self, exc_type, exc_value, traceback):
-        return self.sync(self.__aexit__, exc_type, exc_value, traceback)
+        aw = self.close()
+        assert aw is None
```
Comment (gjoseph92) [inline distributed/deploy/cluster.py:527]:
```suggestion
        assert aw is None, aw
```

Comment (gjoseph92) [reply distributed/deploy/cluster.py:527]: `__exit__` shouldn't be used if `self.asynchronous` is True, right?

File: distributed/utils.py [hunk 2]
```diff
@@ -407,6 +407,61 @@ def wait(timeout):
...
+            return self._event is not None and self._event.is_set()
+
+        async def wait(self) -> bool:
+            if self._event is None:
+                self._event = asyncio.Event()
+
+            return await self._event.wait()
+
+
+class _CollectErrorThread:
```
Comment (gjoseph92) [inline distributed/utils.py:438]: Is this copied from somewhere?

File: distributed/utils.py [hunk 3]
```diff
@@ -470,62 +518,53 @@ def start(self):
...
+                    raise
+                start_exc = e
+                start_evt.set()
+
+        self._loop_thread = _CollectErrorThread(
+            target=run_loop, daemon=True, name="IO loop"
+        )
+        start_evt.wait(timeout=10)
+        if start_exc is not None:
+            raise start_exc
```
Comment (gjoseph92) [inline distributed/utils.py:563]: I feel like having test coverage for this would be good

File: distributed/utils.py [hunk 4]
```diff
@@ -470,62 +518,53 @@ def start(self):
...
-                    raise TypeError(
-                        f"not an exception: {start_exc[0]!r}",
-                    )
-                raise start_exc[0]
-            self._all_loops[self._loop] = count + 1, None
-        else:
-            assert start_exc[0] is None, start_exc
-            self._loop_thread = thread
-            self._all_loops[self._loop] = count + 1, self
+        self._all_loops[loop] = (1, self)
```
Comment (gjoseph92) [inline distributed/utils.py:567]: I think this is an absurd case to worry about, but is it technically possible to somehow have multiple `LoopRunner`s starting in parallel in multiple threads, but targeting the same `loop`, and then overwriting each other here?

File: distributed/utils.py [hunk 5]
```diff
@@ -470,62 +518,53 @@ def start(self):
...
-        if self._asynchronous or real_runner is not None or count > 0:
+        if self._loop is not None:
+            try:
+                count, real_runner = self._all_loops[self._loop]
+            except KeyError:
+                assert self._loop.asyncio_loop.is_running()  # type: ignore[attr-defined]
+                self._started = True
+                return
+
             self._all_loops[self._loop] = count + 1, real_runner
```
Comment (gjoseph92) [inline distributed/utils.py:532]: If multiple `LoopRunner`s are spawned in separate threads, but all given the same loop, this mutation isn't thread-safe. (IDK if `LoopRunner` is supposed to be thread-safe in the first place.)

File: distributed/utils.py [hunk 6]
```diff
@@ -541,23 +580,25 @@ def _stop_unlocked(self, timeout):
...
-                real_runner._real_stop(timeout)
+            return
+
+        assert count == 1
+        del self._all_loops[self._loop]
+        real_runner._real_stop(timeout)
 
     def _real_stop(self, timeout):
         assert self._loop_thread is not None
         if self._loop_thread is not None:
```
Comment (gjoseph92) [inline distributed/utils.py:598]: the assertion followed by conditional feels unnecessary

**Review by gjoseph92 — COMMENTED** [review]
Some questions about thread-safety, but this seems good to me overall.

File: distributed/utils.py (hunk 2 above)
Comment (graingert) [reply distributed/utils.py:438]: It's not copied from anywhere

File: distributed/utils.py (hunk 5 above)
Comment (graingert) [reply distributed/utils.py:532]: This is run protected by a lock

File: distributed/utils.py (hunk 4 above)
Comment (graingert) [reply distributed/utils.py:567]: This is run protected by a lock

File: distributed/utils.py [hunk 7]
```diff
@@ -420,6 +420,61 @@ def wait(timeout):
...
+    from asyncio import Event as LateLoopEvent
+else:
+    # In python 3.10 asyncio.Lock and other primitives no longer support
+    # passing a loop kwarg to bind to a loop running in another thread
+    # e.g. calling from Client(asynchronous=False). Instead the loop is bound
+    # as late as possible: when calling any methods that wait on or wake
+    # Future instances. See: https://bugs.python.org/issue42392
+    class LateLoopEvent:
+        def __init__(self) -> None:
+            self._event: asyncio.Event | None = None
```
Comment (crusaderky) [inline distributed/utils.py:433]:
```suggestion
        _event: asyncio.Event | None

        def __init__(self) -> None:
            self._event = None
```

File: distributed/utils.py [hunk 8]
```diff
@@ -443,35 +498,28 @@ class LoopRunner:
...
-                    stacklevel=2,
-                )
-            self._loop = loop
+                except RuntimeError as e:
+                    raise RuntimeError(
+                        "Constructing LoopRunner(asynchronous=True) without a running loop is not supported"
+                    ) from e
+                loop = IOLoop.current()
+        elif not loop.asyncio_loop.is_running():  # type: ignore[attr-defined]
+            # LoopRunner is not responsible for starting a foriegn IOLoop
```
Comment (crusaderky) [inline distributed/utils.py:513]:
```suggestion
            # LoopRunner is not responsible for starting a foreign IOLoop
```

File: distributed/utils.py [hunk 9]
```diff
@@ -443,35 +498,28 @@ class LoopRunner:
     ] = weakref.WeakKeyDictionary()
     _lock = threading.Lock()
```
Comment (crusaderky) [inline distributed/utils.py:499]:
```suggestion
    _lock = threading.Lock()
    _loop_thread: _CollectErrorThread | None
```

File: distributed/utils.py [hunk 10]
```diff
@@ -443,35 +498,28 @@ class LoopRunner:
...
+        elif not loop.asyncio_loop.is_running():  # type: ignore[attr-defined]
+            # LoopRunner is not responsible for starting a foriegn IOLoop
+            raise RuntimeError(
+                "Constructing LoopRunner(loop=loop) without a running loop is not supported"
+            )
+
+        self._loop = loop
         self._asynchronous = asynchronous
-        self._loop_thread = None
+        self._loop_thread: _CollectErrorThread | None = None
```
Comment (crusaderky) [inline distributed/utils.py:520]:
```suggestion
        self._loop_thread = None
```

File: distributed/deploy/cluster.py [hunk 11]
```diff
@@ -5,6 +5,7 @@
 import logging
 import uuid
 import warnings
+from collections.abc import Awaitable
```
Comment (graingert) [inline distributed/deploy/cluster.py:8]:
```suggestion
```

**Review by crusaderky — APPROVED** [review]

**Pull request merged**